auto-apt.1
.\"                                      Hey, EMACS: -*- nroff -*-
.\" First parameter, NAME, should be all caps
.\" Second parameter, SECTION, should be 1-8, maybe w/ subsection
.\" other parameters are allowed: see man(7), man(1)
.TH auto-apt 1 "29 Jun 2000"
.\" Please adjust this date whenever revising the manpage.
.\"
.\" Some roff macros, for reference:
.\" .nh        disable hyphenation
.\" .hy        enable hyphenation
.\" .ad l      left justify
.\" .ad b      justify to both left and right margins
.\" .nf        disable filling
.\" .fi        enable filling
.\" .br        insert line break
.\" .sp <n>    insert n+1 empty lines
.\" for manpage-specific macros, see man(7)
.SH NAME
auto-apt \- on demand package installation tool
.SH SYNOPSIS
.B auto-apt
.RI [ options ] " " [ command ] " " [ args... ]
.SH DESCRIPTION
This manual page documents briefly the
.B auto-apt
commands.
.PP
\fBauto-apt\fP is a program that checks file access of programs running
within \fBauto-apt\fP environments.  If a program will access a file of
uninstalled package, \fBauto-apt\fP will install the package containing
the file, by using apt-get.
.SH OPTIONS
A summary of options is included below.
.TP
.B "-h"
Usage.
.TP
.B "-s"
No action, perform a simulation of action.
.TP
.B "-y"
Automatic yes to prompt.
.TP
.B "-q"
Quiet
.TP
.B "-i"
Don't install in background process.
.TP
.B "-x"
Disable X interface.
.TP
.B "-X"
Enable X interface. Run apt-get in another \fBx-terminal-emulator\fP 
if you run in X environments.
.TP
.BI "-a [" dist [, dist... ]]
Select installable distribution from 
.BR main , contrib , non-free , non-US
or
.B none.
If you set "-a none", it means almost same as "-s".
.TP
.BI "-p [" hook [, hook... ]]
Select hook functions from
.BR exec , open , access , stat
or
.B none
If you set "-p none", it means almost same as "-s".
.TP
.BI "-D [" dbfile ]
Change \fIdbfile\fP from default 
.B /var/cache/auto-apt/pkgcontents.db
.TP
.BI "-F [" filedb ]
Change \fIfiledb\fP from default
.B /var/cache/auto-apt/pkgfiles.db,
which is generated by \fBauto-apt update-local\fP commands.
It is used as default \fIdbfile\fP for commands with \fB-local\fP 
postfix.  In many cases, it is used with 
.B -L
option to detect file access
.TP
.BI "-L [" record.lists ]
Append file access data to \fIrecord.lists\fP file.

.SH COMMANDS
A summary of command is included below. 
.TP
.BI "run  [" command ]
Run command controlled by \fPauto-apt\fP.  If \fIcommand\fP is not 
specified, user's shell will be used.  
.TP
.B "update [" -a ]
Recreate \fIdbfile\fP from sources.list information.
\fIdbfile\fP can be specified by \fB-D\fP option.
With \fB-a\fP option, all files are stored in \fIdbfile\fP.
Note that \fIdbfile\fP become too big with \fB-a\fP option.
.TP
.B "update-local [" -a ]
Generate \fIfiledb\fP from installed package information.  
\fIfiledb\fP can be specified by \fB-F\fP option.
With \fB-a\fP option, all files are stored in \fIdbfile\fP.
Note that \fIdbfile\fP become too big with \fB-a\fP option.
.TP
.B "merge [" -a ]
Merge package,filename information to \fIdbfile\fP, which can be
specified by \fB-D\fP option.
With \fB-a\fP option, all files are stored in \fIdbfile\fP.
Note that \fIdbfile\fP become too big with \fB-a\fP option.
.TP
.BI del " package "
Delete file information of \fIpackage\fP from \fIdbfile\fP database.
.TP
.BI check " " "[-v] [-f] " file
Check which package provides \fIfile\fP, from \fIdbfile\fP database.
options: \fB-v\fP means verbose. \fB-f\fP uses Contents-*.gz file
instead of \fIdbfile\fP database.
.TP
.BI install " " "[-s] [-d] [-u] [-y] [-v] " file
Try to install a package that provides \fIfile\fP by using apt-get.
You need root privilege to run apt-get install.
.TP
.B list [-v] [-f]
List file information in \fIdbfile\fP database.
options: \fB-v\fP means verbose. \fB-f\fP uses Contents-*.gz file
instead of \fIdbfile\fP database.
.TP
.BI search " " "[-v] [-f] " pattern
Search file information by 
.BR grep (1)
\fIpattern\fP in \fIdbfile\fP database.
options: \fB-v\fP means verbose. \fB-f\fP uses Contents-*.gz file
instead of \fIdbfile\fP database.
.TP
.BI "debuild [" debuildopts ]
Run 
.BR debuild (1)
in auto-apt environments, and show \fBBuild-Depends:\fP estimation.
Recommended to use with \fB-s\fP option of auto-apt.
.TP
.B status
Show whether current environment is in \fBauto-apt\fP or not.
.\" .SH SEE ALSO
.\" .BR auto-apt-pkgcdb (8)
.br
.SH FILES
.IP o
/etc/auto-apt/paths\&.list
path lists
.IP o
/var/cache/auto-apt
cache/database directory
.IP o
/etc/auto-apt/sources.list
sources.list for auto-apt. if not found, symlink to apt-get's sources.list
will be created.
.\" .IP
.\" .IP o
.\" /etc/auto-apt/commands\&.list
.\" command configuration (not yet implemented)
.SH ENVIRONMENTS
.IP o
AUTO_APT_CACHEDIR
override default cache/database directory
.SH BUGS
Known bugs. See 
.B /usr/share/doc/auto-apt/BUGS.
and/or http://bugs.debian.org/auto-apt
.SH AUTHOR
This program was written by Fumitoshi UKAI <ukai@debian.or.jp>,
inspired by idea/comments from Debian Guru (debian.gr.jp) and 
Project Dice members.


debian/changelog
auto-apt (0.3.24) unstable; urgency=medium

  * QA upload.
  * Patches supplied by Bruno Nova, with thanks.
  * Plug a memory leak caused by not freeing lines read from stdin. (LP:
    #1300987)
  * Use dpkg --print-architecture instead of deprecated dpkg
    --print-installation-architecture (the latter causes a warning). (LP:
    #1309991)
  * Accept no leading spaces at the start of a line in sources.list. (LP:
    #1309991)

 -- Iain Lane <iain@orangesquash.org.uk>  Thu, 01 May 2014 10:31:33 +0100

auto-apt (0.3.23) unstable; urgency=low

  * QA upload.
  * Replace PATH_MAX to build on GNU/Hurd. (Closes: #659025).
    - Thanks to Tanguy Le Carrour for the patch.
  * Allow spaces in sources.list. (Closes: #667052).
    - Thanks to Mateusz Poszwa for the patch.
  * Change debian/rules to utilize dh 7.
  * Add debian/manpages file.
  * Bump Standards Version to 3.9.4.

 -- Barry deFreese <bdefreese@debian.org>  Mon, 13 May 2013 18:31:46 -0400

auto-apt (0.3.22) unstable; urgency=low

  * QA upload.
    - ACK NMU.
    - Move to packaging format "3.0 (native)".
  * debian/compat
    - Update to 7.
  * debian/control
    - (Maintainer): Set to Debian QA Group. Package orphaned
      in Bug#541016.
    - (Build-Depends): update to debhelper 7.1. Add quilt.
    - (Depends): add ${misc:Depends}.
    - (Standards-Version): update to 3.8.4
  * debian/patches
    - (Number 10): Fix FTBFS(hurd-i386): PATH_MAX undeclared
      (Closes: #257087).
    - (Number 20): Fix conflicting wget(1) options (Closes: #481658).
      Patch thanks to Krzysztof Burghardt <krzysztof@burghardt.pl>
  * debian/README.source
    - New file.
  * debian/rules
    - Install quilt.
    - (install): use dh_prep. Adjust Makefile call.
  * debian/source/format
    - New file.

 -- Jari Aalto <jari.aalto@cante.net>  Thu, 01 Apr 2010 07:25:46 +0300

auto-apt (0.3.21+nmu1) unstable; urgency=low

  * Non-maintainer upload.
  * Move devscripts from Build-Depends to Recommends. It's not used
    in the build process but in the program itself.
    (Closes: #475259)

 -- Thijs Kinkhorst <thijs@debian.org>  Sat, 03 May 2008 18:04:57 +0200

auto-apt (0.3.21) unstable; urgency=low

  * debian/postrm; remove cache file when purged.
    closes: Bug#275579
  * auto-apt.sh: remove XSI:ism
    closes: Bug#259289, Bug#375931
  * auto-apt.1: fix typos
    closes: Bug#307025, Bug#275278
  * apply patch to auto-apt.es.1
    closes: Bug#257829
  * add auto-apt.fr.1
    closes: Bug#277535
  * debian/control: fix typo in description
    closes: Bug#243994
  * debian/{compat,rules}: debhelper compat level to 4

 -- Fumitoshi UKAI <ukai@debian.or.jp>  Sun, 23 Jul 2006 09:20:01 +0900

auto-apt (0.3.20) unstable; urgency=low

  * auto-apt.sh: add install command
  * paths.list: add /usr/lib/pkgconfig/
  * auto-apt.es.1: added.  spanish manpage
    closes: Bug#206202
  * auto-apt.sh: see /etc/auto-apt/sources.list instead /etc/apt/sources.list.
     if not found, this file is symlinked to /etc/apt/sources.list.
     if you want to have different sources.list for auto-apt than apt-get,
     you can create /etc/auto-apt/sources.list.
    closes: Bug#202979

 -- Fumitoshi UKAI <ukai@debian.or.jp>  Sat, 20 Mar 2004 02:50:29 +0900

auto-apt (0.3.19) unstable; urgency=low

  * fix FTBFS with gcc-3.3. closes: Bug#195085
  - auto-apt-pkg.c (usage): don't use multiline string literals
  - auto-apt-pkg.c: include <time.h> for time()
  - pkgcdb/mempool.c (mempool_restore): fix warnings: 
  	concatenation of string literals with __FUNCTION__ is deprecated

 -- Fumitoshi UKAI <ukai@debian.or.jp>  Mon,  2 Jun 2003 01:28:46 +0900

auto-apt (0.3.18) unstable; urgency=low

  * use --timestamping --continue option for wget
    closes: Bug#187658

 -- Fumitoshi UKAI <ukai@debian.or.jp>  Sun,  6 Apr 2003 02:29:39 +0900

auto-apt (0.3.17) unstable; urgency=low

  * show progress bar while downloading Contents.gz by wget
    closes: Bug#146178, Bug#162068

 -- Fumitoshi UKAI <ukai@debian.or.jp>  Thu, 26 Sep 2002 01:40:38 +0900

auto-apt (0.3.16) unstable; urgency=low

  * fix RE bug reading sources.list, closes: Bug#160662
  * fix obscur message for a search command when db is not initialised
    closes: Bug#146177

 -- Fumitoshi UKAI <ukai@debian.or.jp>  Wed, 25 Sep 2002 00:28:08 +0900

auto-apt (0.3.15) unstable; urgency=low

  * The "Sympathy For The Devil" release
  * add support rsh|ssh method, closes: Bug#94105
  * fix typo in examples/auto-apt-get-menu.sh, closes: Bug#127536

 -- Fumitoshi UKAI <ukai@debian.or.jp>  Sun,  6 Jan 2002 03:41:37 +0900

auto-apt (0.3.14) unstable; urgency=low

  * remove apt, sudo and perl dependency from Depends:
    move these to Recommends:.
    This is because these are only required when user uses auto-apt as
    automatic installation tool.   For users who want to search package
    by file, these packages are not requires.
    See also Bug#125386: ITP: apt-file

 -- Fumitoshi UKAI <ukai@debian.or.jp>  Thu, 20 Dec 2001 01:51:37 +0900

auto-apt (0.3.13) unstable; urgency=low

  * use $(CC) instead of $(LD)
    closes: Bug#105593

 -- Fumitoshi UKAI <ukai@debian.or.jp>  Wed, 18 Jul 2001 00:13:07 +0900

auto-apt (0.3.12) unstable; urgency=low

  * The "Ballad of Fallen Angels" release
  * fix requirepkg() that auto-apt debuild want to 
    apt-get 'build-essential' even if installed
    closes: Bug#101120
  * update paths.list

 -- Fumitoshi UKAI <ukai@debian.or.jp>  Tue, 19 Jun 2001 23:19:24 +0900

auto-apt (0.3.11) unstable; urgency=low

  * The "Gateway Shuffle" release
  * adapt apt-config change in apt 0.5
    fix apt sources.list:etc/apt//SOURCES_LIST not found
    (thanks <taru@debian.org>)

 -- Fumitoshi UKAI <ukai@debian.or.jp>  Tue, 27 Feb 2001 20:27:52 +0900

auto-apt (0.3.10) unstable; urgency=low

  * The "Honky Tonk Women" release
  * fix perl dependency, closes: Bug#80689

 -- Fumitoshi UKAI <ukai@debian.or.jp>  Thu, 25 Jan 2001 13:00:34 +0900

auto-apt (0.3.9) unstable; urgency=low

  * The "Stray Dog Strut" release

auto-apt.sh
#!/bin/sh
# auto-apt -- on demand package installation tool
# Copyright (c) 2000 Fumitoshi UKAI <ukai@debian.or.jp>
# GPL
# 
# AUTO_APT_DEBUG=<defined?>
# AUTO_APT_VERBOSE=<defined?>
# AUTO_APT_QUIET=<defined?>
# AUTO_APT_LOG=<filename>	-- obsolate?
# AUTO_APT_ACCEPT={main,non-US,non-free,contrib,none}
# AUTO_APT_SIMULATE=<defined?>
# AUTO_APT_CACHEDIR=/var/cache/auto-apt/
# AUTO_APT_DETECT=<dbfile>
# AUTO_APT_YES=<defined?>
# AUTO_APT_DB=/var/cache/auto-apt/pkgcontents.bin
# AUTO_APT_HOOK={exec,open,access,stat}
#
# APT_CONFIG=/etc/apt/apt.conf
# SOURCES_LIST=/etc/auto-apt/sources.list (cf: apt-config -dump)

prog=`basename $0`

AUTO_APT_ID='$Id: auto-apt.sh,v 1.58 2004/03/19 17:53:25 ukai Exp $'
AUTO_APT_PKGCDB=${AUTO_APT_PKGCDB:-'/usr/lib/auto-apt/auto-apt-pkgcdb'}
AUTO_APT_SO=${AUTO_APT_SO:-/lib/auto-apt.so}
AUTO_APT_ACCEPT=${AUTO_APT_ACCEPT:-"main,non-US,non-free,contrib"}
AUTO_APT_CACHEDIR=${AUTO_APT_CACHEDIR:-"/var/cache/auto-apt"}
AUTO_APT_DB=${AUTO_APT_DB:-"${AUTO_APT_CACHEDIR}/pkgcontents.bin"}
AUTO_APT_FILEDB=${AUTO_APT_FILEDB:-"${AUTO_APT_CACHEDIR}/pkgfiles.bin"}
AUTO_APT_HOOK=${AUTO_APT_HOOK:-"all"}
AUTO_APT_SOURCES_LIST=${AUTO_APT_SOURCES_LIST:-"/etc/auto-apt/sources.list"}

export AUTO_APT_ACCEPT
export AUTO_APT_HOOK

LIBC_PATH=`ldd /bin/sh | grep libc.so | awk '{print $3}'`
if test -f $LIBC_PATH; then
   export LIBC_PATH
fi

requirepkg() {
  cmd=$1
  pkg=$2
  if test -e "$1";then
    :
  else
    echo "I: Install:$2	file:$1	by:$0" >&2
    apt-get -y install $2 >&2
  fi
}

usage() {
  echo "$AUTO_APT_ID
Usage: auto-apt [options] command [arg ...]

auto-apt is a simple command line interface for setting up auto-apt
environment and/or search packages by filename.

Commands:
    run - Enter auto-apt environment
	run [command [cmdarg]]
    update - Retrieve new lists of Contents (available file list)
	update
    updatedb - Regenerate lists of Contents (available file list, no download)
        updatedb
    update-local - Generate installed file lists
    	update-local
    merge - Merge lists of Contents
	merge
    del - Delete package list
	del package
    check - Check which package will provide the filename
        check [-v] [-f] filename
    list - List filelist in dbfile
    	list [-v] [-f]
    search - Search package by filename (grep)
        search [-v] [-f] pattern
    debuild - debuild with auto-apt
	useful to get build-depends:
    status - Report current environments (auto-apt or not)

 For some commands, command name with "-local" suffix, it will use 
 pkgfiles.db, which is created by update-local command, instead of
 pkgcontents.db

Options:
    [-s] [-y] [-q] [-i] [-x] [-X]
    [-a dists] [-p hooks]
    [-D pkgcontents.bin] [-F pkgfiles.bin] [-L detect.list] 

See the auto-apt(1) manual pages.
" >&2
    exit
}

run() {
    if test -e /usr/bin/apt-get; then
	:
    else
	echo E: You need apt-get to run auto-apt.
	exit 1
    fi
    if test -e /usr/bin/sudo; then
	:
    else
	echo W: without sudo, you can not install new packages automatically.
    fi
    if test "$1" = ""; then
	set -- $SHELL
    fi
    if test "$AUTO_APT_QUIET" = ""; then
      echo Entering auto-apt mode: "$@"
      echo Exit the command to leave auto-apt mode.
    fi
    command -v "$1" |grep -q / > /dev/null 2>&1 && LD_PRELOAD=$AUTO_APT_SO:$LD_PRELOAD exec "$@"
    echo "E: Exec $1 failed, auto-apt failed"
}

update() {
    if test "$1" = db; then
       nodownload=1
       shift
    fi
    # dirname in shellutils, which is required
    CACHEDIR=`dirname $AUTO_APT_DB`
    test -d "$CACHEDIR" || {
      echo "E: update: cache dir not found: $CACHEDIR" >&2; exit
    }
    test -w "$CACHEDIR" || {
      echo "E: You need write permission for $CACHEDIR" >&2; exit
    }
    ARCH=$(dpkg --print-architecture)

    cd "$CACHEDIR" && {
      rm -f "$AUTO_APT_DB"

      test -f $AUTO_APT_SOURCES_LIST || {
        eval `apt-config shell APT_ETC Dir::Etc/ SOURCES_LIST Dir::Etc::sourcelist`
        test -f $APT_ETC/$SOURCES_LIST || {
          echo "E: apt sources.list:$APT_ETC/$SOURCES_LIST not found" >&2; exit
        }
	ln -sf $APT_ETC/$SOURCES_LIST $AUTO_APT_SOURCES_LIST
      }
      cat $AUTO_APT_SOURCES_LIST | \
	sed -ne 's=^[[:space:]]*deb[[:space:]][[:space:]]*\([^:[:space:]]*\):\([^[:space:]]*\)[[:space:]][[:space:]]*\([^[:space:]]*\)[[:space:]][[:space:]]*.*=\1 \2/dists/\3=p' | 
      while read method url
      do
         u=`echo $url | sed -e 's,^/*,,' -e 's:/:_:g' -e 's:__*:_:g'`
	 case "$method" in
	 http|ftp)
	   if test "$nodownload" = 1; then
	      :
	   else
  	     echo "Downloading $method:$url Contents-$ARCH.gz ..." >&2
	     # proxy setting
	     if test "${method}_proxy" = ""; then
                host=`echo $url | sed -e 's=^//==' -e 's=/.*=='`
		eval `apt-config shell ${method}_proxy Acquire::${method}::Proxy`
		eval `apt-config shell ${method}_proxy Acquire::${method}::Proxy::${host}`
		if test "${method}_proxy" = "DIRECT"; then
	            unset ${method}_proxy
		else
		    export ${method}_proxy
	        fi
	     fi
	     # other options
	     wgetopt="--timestamping --continue --progress=bar:force"
	     if test "$method" = "ftp"; then
		passive_ftp="true"
		eval `apt-config shell passive_ftp Acquire::ftp::Passive`
		if test "$passive_ftp" = "true"; then
		    wgetopt="$wgetopt --passive-ftp"
		fi
	     fi
	     requirepkg /usr/bin/wget wget
	     # /usr/bin/wget -q -O ${u}_Contents-$ARCH.gz $method:$url/Contents-$ARCH.gz >&2
             wget $wgetopt -O ${u}_Contents-$ARCH.gz $method:$url/Contents-$ARCH.gz 2>&1 | tail -n +8 >&2
           fi
	   ;;
	 rsh|ssh)
	   if test "$nodownload" = 1; then
	      :
	   else
  	     echo -n "Downloading $method:$url Contents-$ARCH.gz ..." >&2
	     case "$method" in
	     rsh) requirepkg /usr/bin/rsh rsh-client; cmd=rcp;;
	     ssh) requirepkg /usr/bin/ssh ssh; cmd=scp;;
	     esac
	     path=`echo $url | sed -e 's=//\([^/]*\)/\(.*\)$=\1:/\2='`
	     $cmd $path/Contents-$ARCH.gz ${u}_Contents-$ARCH.gz > /dev/null 2>&1
	     echo "done" >&2
           fi
	   ;;
	 file)
           if test "$nodownload" = 1; then
	       :
	   else
	     ln -fs $url/Contents-$ARCH.gz ${u}_Contents-$ARCH.gz
	   fi
	   ;;
	   
	 *)
	   echo "W: Unsupported apt-line: $method, ignored" >&2
	   ;;
	 esac
	 test -s ${u}_Contents-$ARCH.gz &&
	 gunzip < ${u}_Contents-$ARCH.gz |
	 sed -e '1,/^FILE/d'
	 echo "" >&2
      done | $AUTO_APT_PKGCDB -f "$AUTO_APT_DB" put "$@"
    }
}

updatelocal() {
    test -f "$AUTO_APT_FILEDB" && rm -f "$AUTO_APT_FILEDB"
    ls -1rt /var/lib/dpkg/info/*.list | while read list
    do
      pkg=`echo $list | sed -ne 's:.*/dpkg/info/\(.*\).list:\1:p'`
      awk -v pkg=$pkg '{print $1, " ", pkg}' $list
    done | $AUTO_APT_PKGCDB -f "$AUTO_APT_FILEDB" put "$@"
}

merge() {
    $AUTO_APT_PKGCDB -f "$AUTO_APT_DB" put "$@"
}

del() {
    if test "$1" = ""; then
        echo "E: del: missing packagename" >&2
	exit
    fi
    $AUTO_APT_PKGCDB -f "$AUTO_APT_DB" del "$@"
}

check() {
    while getopts "fv" opt; do
    case "$opt" in
     f) full=yes;;
     v) verbose=yes;;
     *) echo "$0 check [-f] filename" >&2; exit;;
    esac
    done
    shift $(($OPTIND -1))
    if test "$1" = ""; then
    	echo "E: check: missing filename" >&2
	exit
    fi
    if test "$full" = "yes"; then
      pat=`echo $1 | sed -e 's:^\/::'`
      (cd `dirname $AUTO_APT_DB`
        for contents in *_Contents-*.gz
        do
          test -s $contents || continue
          gunzip < $contents | sed -e '1,/^FILE/d'
        done ) | grep -E "^/?$pat[[:space:]]" | awk '{
	    print $2;
	    if ("'$verbose'" == "yes") { print $1; } 
        }'
    else
      pkgcdbopt=
      test "$verbose" = "yes" && pkgcdbopt="-v"
      $AUTO_APT_PKGCDB -f "$AUTO_APT_DB" get $pkgcdbopt "$@"
    fi
}

install() {
    while getopts "sduyv" opt; do
     case "$opt" in
      s) simulate=yes;;
      d) download=yes;;
      u) upgraded=yes;;
      y) assumeyes=yes;; 
      v) verbose=yes;;
      *) echo "$0 install [-s] [-d] [-u] [-y] [-v] filename" >&2; exit;;
     esac
    done
    shift $(($OPTIND -1))
    filename="$1"
    test "$verbose" = "yes" && echo "N: try to install a package providing $filename"
    pkgs=$($0 check "$filename")
    IFS0="$IFS"; IFS=,; set -- $pkgs; IFS="$IFS0"
    test "$verbose" = "yes" && echo "N: $@"
    if [ $# -eq 0 ]; then
       echo "E: no candidate found for $filename" >&2; exit 1
    fi
    if [ $# -gt 1 ]; then
       echo "E: multiple candidates found for $filename" >&2
       echo "I:  $@" >&2
       exit 1
    fi
    pkgname=${1##*/}
    aptopt=
    test "$simulate" = "yes" && aptopt="$aptopt -s"
    test "$download" = "yes" && aptopt="$aptopt -d"
    test "$upgraded" = "yes" && aptopt="$aptopt -u"
    test "$assumeyes" = "yes" && aptopt="$aptopt -y"
    test "$verbose" = "yes" && echo "apt-get $aptopt intall $pkgname"
    apt-get $aptopt install $pkgname
}

list() {
    while getopts "fv" opt; do
     case "$opt" in
      f) full=yes;;
      v) verbose=yes;;
      *) echo "$0 list [-f]" >&2; exit;;
     esac
    done
    shift $(($OPTIND -1))
    if test "$verbose" = "yes"; then
       set -- -v "$@"
    fi
    if test "$full" = "yes"; then
      (cd `dirname $AUTO_APT_DB`
        founddb=false
        for contents in *_Contents-*.gz
        do
          test -s $contents || continue
          founddb=true
          gunzip < $contents | sed -e '1,/^FILE/d'
        done
	if test "$founddb" = false; then
	   echo "E: no auto-apt db, run 'auto-apt update' first"
	   exit 1
	fi
       )
    else
      if test -f "$AUTO_APT_DB"; then
        $AUTO_APT_PKGCDB -f "$AUTO_APT_DB" list
      else
	echo "E: $AUTO_APT_DB not found, run 'auto-apt update' first"
	exit 1
      fi
    fi
}

search() {
    if test "$1" = ""; then
       echo "E: search: missing search pattern" >&2
       exit
    fi
    while getopts "fv" opt; do
     case "$opt" in
      f) full=yes;;
      v) verbose=yes;;
      *) echo "$0 search [-f] pattern" >&2; exit;;
     esac
    done
    shift $(($OPTIND -1))
    pat=$1
    if test "$verbose" = "yes"; then
       set -- -v
    fi
    if test "$full" = "yes"; then
       set -- -f "$@"
    fi
    list "$@" | grep "$pat"
}

debuild() {
  test -f $AUTO_APT_FILEDB || {
    if test "$AUTO_APT_YES" = ""; then
     fdb=`basename $AUTO_APT_FILEDB`
     echo "You have no $AUTO_APT_FILEDB"
     echo "It is recommended to create $fdb to get right build-depends, "
     echo "but it takes too much time to generate $fdb."
     echo -n "Create it now ? [Y/n] "
     while read yn;
     do 
       test "$yn" = "" && yn=y
       case "$yn" in
        Y|y) auto-apt update-local; break;;
        N|n) echo "ok, run without $fdb"; break;;
        *) echo -n "Create it now ? [Y/n] "; continue;;
      esac
     done
    else
     auto-apt update-local
    fi
  }
  if test "$AUTO_APT_DETECT" = ""; then
   requirepkg /usr/bin/dpkg-parsechangelog dpkg-dev
   package=`dpkg-parsechangelog | sed -n 's/^Source: //p'`
   test "$package" = "" && {
     echo "E: badly formatted changelog file, no Source:?" >&2; exit
   }
   version=`dpkg-parsechangelog | sed -n 's/^Version: //p'`
   test "$version" = "" && {
     echo "E: badly formatted changelog file, no Version:?" >&2; exit
   }
   # dpkg-architecture is dpkg-dev
   arch=`dpkg-architecture -qDEB_HOST_ARCH 2>/dev/null` \
     && test "${arch}" != "" \
     || arch=`dpkg --print-architecture`
   sversion=`echo "$version"| perl -pe 's/^\d+://'`
   pva="${package}_${sversion}${arch:+_${arch}}"
   DDB=../${pva}.lists
   AUTO_APT_DETECT=`pwd`/$DDB
   export AUTO_APT_DETECT
  else
   DDB=$AUTO_APT_DETECT
  fi
  test -f $AUTO_APT_DETECT && {
   if test "$AUTO_APT_YES" = ""; then
    echo "detect.lists $AUTO_APT_DETECT already existed"
    echo -n "Remove it, before running debuild? [Y/n] "
    while read yn;
    do
      test "$yn" = "" && yn=y
      case "$yn" in
       Y|y) rm -f $AUTO_APT_DETECT; break;;
       N|n) echo "ok, append record to $AUTO_APT_DETECT"; break;;
       *) echo -n "Remove it, before running debuild? [Y/n]"; continue;;
      esac
    done
   else
    rm -f $AUTO_APT_DETECT
   fi
  }
  rm -f $AUTO_APT_DETECT
  echo "$prog: record file access to $DDB"
  requirepkg /usr/bin/debuild devscripts
  auto-apt -L $AUTO_APT_DETECT run /usr/bin/debuild -E "$@"
  echo "Build dependencies will be ... (in ../${pva}.build-depends)"
  rm -f ../essential-packages-list
  requirepkg /usr/share/doc/build-essential/essential-packages-list build-essential
  sed -e '1,/^$/d' /usr/share/doc/build-essential/essential-packages-list \
    > ../essential-packages-list
  cat $DDB | sed -ne '/^\//p' | awk '$2 !~ /[*!]/ {print $2}' | 
    grep -Fvf ../essential-packages-list | sort | uniq | \
    perl -e 'print "  " . join(", ", grep {chomp} <STDIN>) . "\n"' \
	> ../${pva}.build-depends
  rm -f ../essential-packages-list
  cat ../${pva}.build-depends
  echo "Note: file access record is stored $DDB"
}

status() {
    case $LD_PRELOAD in
    /lib/auto-apt.so|/lib/auto-apt.so:*|*:/lib/auto-apt.so|*:/lib/auto-apt.so:*)
     echo "auto-apt mode"; 
     if test "$AUTO_APT_QUIET" != "yes"; then
        test "$AUTO_APT_SIMULATE" = "yes" && echo -n " -s"
	test "$AUTO_APT_YES" = "yes" && echo -n " -y"
	test "$AUTO_APT_NOBG" = "yes" && echo -n " -i"
	test "$AUTO_APT_X" = "yes" && echo -n " -X"
        echo -n " -a $AUTO_APT_ACCEPT";
	echo -n " -p $AUTO_APT_HOOK";
	echo -n " -D $AUTO_APT_DB";
	echo -n " -F $AUTO_APT_FILEDB";
	test -f "$AUTO_APT_DETECT" && echo -n " -L $AUTO_APT_DETECT";
	echo "";
     fi
     exit 0;;
    *)
     echo "normal mode"; exit 1;;
    esac
}

## general command line parsing
# /usr/bin/getopt in util-linux (required)
## XXX: use getopts?

# default -X enable
case X"$AUTO_APT_X" in
Xoff|Xno) ;;
X*) if test "$DISPLAY" != "" && test -x /usr/bin/x-terminal-emulator; then
      AUTO_APT_X=yes; export AUTO_APT_X;
    fi ;;
esac

while getopts "hsyqixXa:c:p:D:F:L:" opt; do
  # echo "opt=$opt OPTIND=$OPTIND OPTARG=$OPTARG"
  case "$opt" in
   h) usage;;
   s) AUTO_APT_SIMULATE=yes; export AUTO_APT_SIMULATE ;;
   y) AUTO_APT_YES=yes; export AUTO_APT_YES;;
   q) AUTO_APT_QUIET=yes; export AUTO_APT_QUIET;;
   i) AUTO_APT_NOBG=yes; export AUTO_APT_NOBG;;
   x) AUTO_APT_X=no; export AUTO_APT_X ;;
   X) if test "$DISPLAY" != "" && test -x /usr/bin/x-terminal-emulator; then
         AUTO_APT_X=yes; export AUTO_APT_X
      else
         echo "no \$DISPLAY or x-terminal-emulator, X support disabled" >&2
      fi;;
   a) case "$OPTARG" in
       none) AUTO_APT_ACCEPT;;
       *) AUTO_APT_ACCEPT="$AUTO_APT_ACCEPT,$OPTARG";;
      esac
      export AUTO_APT_ACCEPT;;
   p) case "$OPTARG" in
       none) AUTO_APT_HOOK;;
       *) AUTO_APT_HOOK="$AUTO_APT_HOOK,$OPTARG";;
      esac
      export AUTO_APT_HOOK;;
   D) AUTO_APT_DB="$OPTARG"; export AUTO_APT_DB;;
   F) AUTO_APT_FILEDB="$OPTARG"; export AUTO_APT_FILEDB ;;
   L) AUTO_APT_DETECT="$OPTARG"; export AUTO_APT_DETECT AUTO_APT_FILEDB ;;
   *) usage;;
  esac
done
# echo "opt=$opt OPTIND=$OPTIND OPTARG=$OPTARG"
shift $(($OPTIND - 1))

if [ $# -lt 1 ]; then
    usage
fi

CMD=$1; shift
case "$CMD" in
 run) run "$@";;
 update) update "$@";;
 updatedb) update db "$@";;
 updatedb-local) AUTO_APT_DB="$AUTO_APT_FILEDB" update "$@" ;;
 update-local) updatelocal "$@";;
 merge) merge "$@";;
 merge-local) AUTO_APT_DB="$AUTO_APT_FILEDB" merge "$@" ;;
 check) check "$@";;
 check-local) AUTO_APT_DB="$AUTO_APT_FILEDB" check "$@" ;;
 install) install "$@";;
 list) list "$@";;
 list-local) AUTO_APT_DB="$AUTO_APT_FILEDB" list "$@" ;;
 del) del "$@";;
 del-local) AUTO_APT_DB="$AUTO_APT_FILEDB" del "$@" ;;
 search) search "$@";;
 search-local) AUTO_APT_DB="$AUTO_APT_FILEDB" search "$@" ;;
 debuild) debuild "$@";;
 status|mode) status "$@";;
 *) echo "unknown command: $CMD" >&2; usage ;;
esac
exit
